src/main/java/com/tabnine/binary/fetch/BootstrapperSupport.java

```diff
@@ -0,0 +1,73 @@
+package com.tabnine.binary.fetch;
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.util.text.SemVer;
+import com.tabnine.general.StaticConfig;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.prefs.Preferences;
+
+import static java.lang.String.format;
+
+public class BootstrapperSupport {
+    static Optional<BinaryVersion> bootstrap_version(LocalBinaryVersions localBinaryVersions, BinaryRemoteSource binaryRemoteSource, BinaryDownloader binaryDownloader) {
+        // bootstrapper can change versions under the hood
+        // plugin must ensure it is running a version containing the bootstrapper but not override any versions that the bootstrapper may have downloaded on its own
+        // we persist the first known bootstrapper version and choose accordingly
+
+        Optional<SemVer> thisVersion = minimalBootstrappedVersion();
+
+        if (thisVersion.isPresent()) {
+            // versions list is returned reverse ordered by semver
+            // take first which has a version >= thisVersion
+            // (since bootstrapper can download newer versions or downgrade but we let the bootstrapper handle that)
+            // "for" since parseFromText returns nullable and don't want to deal with ugliness in map/filter
+            List<BinaryVersion> versions = localBinaryVersions.listExisting();
+            for (BinaryVersion v : versions) {
+                SemVer ver = SemVer.parseFromText(v.getVersion());
+                if (ver.compareTo(thisVersion.get()) >= 0) {
+                    return Optional.of(v);
+                }
+            }
+        }
+
+        // no local bundle exists so we go to the bundle API and get the latest
+        // after that point the bundle will handle the rest
+
+        Optional<BinaryVersion> binaryVersion = binaryRemoteSource
+                                                    .fetchPreferredVersion(StaticConfig.getTabNineBundleVersionUrl())
+                                                    .flatMap(binaryDownloader::downloadBinary);
+
+        if (binaryVersion.isPresent()) {
+            Logger.getInstance(BootstrapperSupport.class).info(format("downloaded bootstrap bundle %s", binaryVersion.get()));
+            String version = binaryVersion.get().getVersion();
+            // we save this as the "minimal bootstrap supported version"
+            get_prefs().put(BOOTSTRAPPED_VERSION_KEY, version);
+        } else {
+            // will fallback to the legacy code in the BinaryVersionFetcher
+            Logger.getInstance(BootstrapperSupport.class).warn("couldn't download bootstrapper bundle!");
+        }
+
+        return binaryVersion;
+
+    }
+
+    static String BOOTSTRAPPED_VERSION_KEY = "bootstrapped version";
+
+    static Preferences get_prefs() {
+        return Preferences.userNodeForPackage(BootstrapperSupport.class);
+    }
+
+    /// returns the minimal known version supporting the bootstrapper
```

Comment:
no need for so much comments. Please remove them all.

Only acceptable comments are when:
* Some optimisation took place, sacrificing readability.
* We took a shortcut and we would like encourage people to improve in the future.